lib/Sema/TypeCheckDeclPrimary.cpp

```diff
@@ -2060,6 +2060,23 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
       break;
     }
     }
+
+    // If the macro has a (non-Void) result type, it must have the freestanding
+    // expression role. Other roles cannot have result types.
+    if (auto resultTypeRepr = MD->getResultTypeRepr()) {
+      if (!MD->getMacroRoles().contains(MacroRole::Expression) &&
```

Comment:
If a macro had multiple roles (including `freestanding(expression)`), should we reject non-void types in that case too?
```swift
@attached(accessor) @freestanding(expression)
macro Foo() -> Int
```
Because in this case the return type `-> Int` still doesn't make sense for the accessor role.

Reply:
No, I don't think we should. The result type only applies to the `expression` role, and it's fine for it to be ignored for any attached roles.

Note that we don't currently diagnose when a macro has multiple freestanding roles. I'm handling that (and related cleanup) via a separate pull request